include/repeater_cadmium.hpp

```diff
@@ -0,0 +1,197 @@
+      * or acknowledgement.
+    */
+    struct state_type {
+        bool ack;
+        bool sending;
+        int packet;
+        int ack_packet;
+    };
+    state_type state;
+
+    /**
+     * Defining input port and output port
+    */
+    using input_ports = std::tuple<typename defs::packet_in,
+        typename defs::ack_in>;
+    using output_ports = std::tuple<typename defs::packet_sent_out,
+        typename defs::ack_received_out>;
+
+    /**
+     * Internal transition function that sets the
+     * state of sending packet and acknowledgement
+     * to passive.
+    */
+    void internal_transition() {
+        state.sending = false;
+        state.ack = false;
+    }
+
+    /**
+     * Function that performs external transition.
+     * Retrieves the messages: if the number of messages
+     * is more than 1, it asserts that only one message is
+     * expected per time unit. After that it determines
+     * from which input port message is coming and based
+     * on the message type sets the state.
+     * @param e time variable
+     * @param mbs message bags
+    */
+    void external_transition(TIME e,
+                             typename make_message_bags<input_ports>::type mbs) {
+        if ((get_messages<typename defs::packet_in>(mbs).size()
+            +get_messages<typename defs::ack_in>(mbs).size()) > 1) {
+            assert(false && "One message at a time");
+        }
+        for (const auto &x : get_messages<typename defs::packet_in>(mbs)) {
+            state.packet = static_cast<int>(x.value);
+            state.sending = true;
+        }
+        for (const auto &x : get_messages<typename defs::ack_in>(mbs)) {
+            state.ack_packet = static_cast<int>(x.value);
+            state.ack = true;
+        }
+    }
+
+    /**
+     * it is the function that make call to internal_transition
+     * followed by call to external transition.
+     * @param e time variable
+     * @param mbs message bags
+    */
+    void confluence_transition(TIME e,
+                               typename make_message_bags<input_ports>::type mbs) {
+        internal_transition();
+        external_transition(TIME(), std::move(mbs));
+    }
+
+    /**
+     * Funtion that sends the packet to the respective
+     * output ports depending on the state type.
+     * When in sending state the value of the input is assigned
+     * to the packet and then sent to the mesage bags.
+     * When in acknowledgement state ack packet value is
+     * assignet ot the output and output is pushed to message bags.
+     * @return Message bags
+    */
+    typename make_message_bags<output_ports>::type output() const {
+        typename make_message_bags<output_ports>::type bags;
+        Message_t out;
+        if (state.sending) {
+            out.value = state.packet;
+            get_messages<typename defs::packet_sent_out>(bags).push_back(out);
+        }
+        else if(state.ack)
```

Comment:
You will need and if instead of an if-else.. It may be the case that you receive at the same time a packet and an acknowledgment in the repeater.

Reply:
This has been fixed.